File: src/main/java/ee/taltech/gtm/listener/CustomVisibleAreaListener.java
package ee.taltech.gtm.listener;

import com.intellij.openapi.editor.event.VisibleAreaEvent;
import com.intellij.openapi.editor.event.VisibleAreaListener;
import com.intellij.openapi.fileEditor.FileDocumentManager;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.vfs.VirtualFile;
import org.jetbrains.annotations.NotNull;

public class CustomVisibleAreaListener implements VisibleAreaListener {

    @Override
    public void visibleAreaChanged(@NotNull VisibleAreaEvent e) {
        FileDocumentManager instance = FileDocumentManager.getInstance();
        VirtualFile file = instance.getFile(e.getEditor().getDocument());
        Project project = e.getEditor().getProject();
        // TODO: dispatch event, or delete this"
    }
}

File: src/main/java/ee/taltech/gtm/listener/CustomVisibleAreaListener.kt
package ee.taltech.gtm.listener

import com.intellij.openapi.editor.event.VisibleAreaEvent
import com.intellij.openapi.editor.event.VisibleAreaListener
import com.intellij.openapi.fileEditor.FileDocumentManager
import ee.taltech.gtm.GtmWrapper

class CustomVisibleAreaListener : VisibleAreaListener {
    override fun visibleAreaChanged(e: VisibleAreaEvent) {
        val instance = FileDocumentManager.getInstance()
        val file = instance.getFile(e.editor.document) ?: return
        val project = e.editor.project ?: return
        GtmWrapper.instance.recordFile(project, file)
    }
}